crates/vim/src/normal.rs

```diff
@@ -327,6 +329,14 @@ fn yank_line(_: &mut Workspace, _: &YankLine, cx: &mut ViewContext<Workspace>) {
     })
 }
 
+fn undo(_: &mut Workspace, _: &Undo, cx: &mut ViewContext<Workspace>) {
+    Vim::update(cx, |vim, cx| {
+        vim.update_active_editor(cx, |_, editor, cx| editor.undo(&editor::actions::Undo, cx));
+        vim.switch_mode(Mode::Visual, false, cx);
+        vim.switch_mode(Mode::Normal, false, cx);
```

Comment:
Setting the mode to visual and then back to normal again is a workaround to a problem that I don't fully understand, but is detailed below.

When adding logging:

```diff
diff --git a/crates/vim/src/normal.rs b/crates/vim/src/normal.rs
index a1f648eef..63b2a386d 100644
--- a/crates/vim/src/normal.rs
+++ b/crates/vim/src/normal.rs
@@ -331,9 +331,11 @@ fn yank_line(_: &mut Workspace, _: &YankLine, cx: &mut ViewContext<Workspace>) {

 fn undo(_: &mut Workspace, _: &Undo, cx: &mut ViewContext<Workspace>) {
     Vim::update(cx, |vim, cx| {
+        log::error!("mode before undo: {:?}", vim.state().mode);
         vim.update_active_editor(cx, |_, editor, cx| editor.undo(&editor::actions::Undo, cx));
-        vim.switch_mode(Mode::Visual, false, cx);
+        log::error!("mode after undo: {:?}", vim.state().mode);
         vim.switch_mode(Mode::Normal, false, cx);
+        log::error!("mode after switch mode to normal: {:?}", vim.state().mode);
     })
 }
```

Which oddly yields `Normal` for each step:

```
[2024-02-25T18:10:16-05:00 ERROR vim::normal] mode before undo: Normal
[2024-02-25T18:10:16-05:00 ERROR vim::normal] mode after undo: Normal
[2024-02-25T18:10:16-05:00 ERROR vim::normal] mode after switch mode to normal: Normal
```

But in reality, the editor's toolbar shows that it's in `--VISUAL LINE--` mode after the undo is complete. The following gif shows me running `shift+v`, `d`, `u`, and then `esc`.

![zed](https://github.com/zed-industries/zed/assets/934497/1bc48c30-99d5-4962-8d41-25b2018cf130)

Reply:
After doing a bit of sleuthing, it appears this behavior comes from https://github.com/zed-industries/zed/blob/53a81bf3f1359037ca7a836cbdc8c8f0d93ba584/crates/vim/src/vim.rs#L662

Which is called whenever selections are changed. Would love any suggestions on how to best handle this.